amlb/datautils.py

```diff
@@ -284,23 +258,97 @@ def inverse_transform(self, vec, **params):
         return self.delegate.inverse_transform(vec, **params)
 
 
-def impute(X_fit, *X_s, missing_values=np.NaN, strategy='mean'):
+def impute_array(X_fit, *X_s, missing_values=np.NaN, strategy="mean"):
     """
-
     :param X_fit:
     :param X_s:
     :param missing_values:
-    :param strategy: 'mean', 'median', 'most_frequent', 'constant' (from sklearn 0.20 only, requires fill_value arg)
+    :param strategy: 'mean', 'median', 'mode', ('constant', value), None
     :return:
     """
-    # TODO: impute only if np.isnan(X_fit).any() ?
-    imputer = Imputer(missing_values=missing_values, strategy=strategy)
-    imputed = imputer.fit_transform(X_fit)
+    if strategy is None:
+        return [X_fit, *X_s]
+    strategy, fill_value = strategy if isinstance(strategy, tuple) and strategy[0] == 'constant' else (strategy, None)
+    strategy = dict(mode='most_frequent').get(strategy, strategy)
+
+    imputer = Imputer(missing_values=missing_values, strategy=strategy, fill_value=fill_value)
+    imputed = _restore_dtypes(imputer.fit_transform(X_fit), X_fit)
     if len(X_s) > 0:
         result = [imputed]
         for X in X_s:
-            result.append(imputer.transform(X))
+            result.append(_restore_dtypes(imputer.transform(X), X))
         return result
     else:
         return imputed
 
+
+def impute_dataframe(X_fit, *X_s, missing_values=np.NaN, strategy='mean'):
+    """
+    :param X_fit:
+    :param X_s:
+    :param missing_values:
+    :param strategy: 'mean', 'median', 'mode', ('constant', value), None
+        or a dictionary specifying a strategy by dtype (int, float, number, category, string, object, datetime)
+    :return:
+    """
+    if strategy is None:
+        return [X_fit, *X_s]
+    print("*** ori ***")
+    print(_rows_with_nas(X_fit))
+    print(X_fit.dtypes)
+    if isinstance(strategy, dict):
+        for dt, s in strategy.items():
+            X_fit.select_dtypes(include=dt)
+    else:
+        imputed = _impute_pd(X_fit, *X_s, missing_values=missing_values, strategy=strategy)
+    print("*** imputed ***")
+    print(imputed[0])
+    print(_rows_with_nas(imputed[0]))
+    print(imputed[0].dtypes)
```

Comment:
oh yes, I have to review those… it's even possible that this function is not used at all: I had issues with encoding when I was trying to pass pandas DF to some frameworks… and I ended up keeping the numpy array for those, so that pandas encoding is not supported at all, so I should just remove this code.